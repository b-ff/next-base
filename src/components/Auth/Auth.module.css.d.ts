declare const styles: {
  readonly "auth": string;
  readonly "profileLink": string;
  readonly "details": string;
  readonly "name": string;
};
export = styles;

