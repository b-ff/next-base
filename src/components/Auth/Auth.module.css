.auth {
  display: inline-flex;
  flex-direction: row;
  align-items: center;
  justify-content: flex-start;
  gap: var(--spacing_2x);
}

.auth a {
  font-size: var(--font_size_small);
  color: var(--color_font_neutral);
}

.details {
  display: flex;
  flex-direction: column;
}

.name {
  font-weight: 500;
  font-size: var(--font_size_regular);
}
