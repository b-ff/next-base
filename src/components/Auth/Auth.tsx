import { FC } from "react";
import { useSession, signIn, signOut } from "next-auth/react";
import Link from "next/link";

import styles from "./Auth.module.css";
import { Avatar } from "../Avatar";
import { Flex } from "../Flex";

export const Auth: FC = () => {
  const { data: session } = useSession();

  return session ? (
    <Flex gap={2} direction="row" className={styles.auth}>
      <Avatar
        src={session.user?.image ?? ""}
        name={session.user?.name ?? ""}
        size={36}
      />
      <Flex gap={0}>
        <strong className={styles.name}>{session.user?.name ?? "â€”"}</strong>
        <Link href="#" onClick={() => signOut()}>
          Sign out
        </Link>
      </Flex>
    </Flex>
  ) : (
    <Link href="#" onClick={() => signIn()}>
      Sign In / Sign Up
    </Link>
  );
};
