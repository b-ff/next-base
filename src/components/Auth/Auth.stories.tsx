import type { Meta, StoryObj } from "@storybook/react";

import { Auth } from "./Auth";
import { SessionContext, SessionContextValue } from "next-auth/react";
import { Session } from "next-auth";

const meta: Meta<typeof Auth> = {
  component: Auth,
  tags: ["autodocs"],
};

export default meta;

type Story = StoryObj<typeof Auth>;

export const SignedIn: Story = {
  args: {},
  render: () => {
    const user = {
      email: "kermit@muppets.com",
      name: "Kermit The Frog",
      image:
        "https://lumiere-a.akamaihd.net/v1/images/character_themuppets_kermit_b77a431b.jpeg",
    };

    const fakeSession = {
      update: () => Promise.resolve({}),
      data: { user, expires: "" },
      status: "authenticated",
    };

    return (
      <SessionContext.Provider value={fakeSession as SessionContextValue}>
        <Auth />
      </SessionContext.Provider>
    );
  },
};

export const SignedOut: Story = {
  args: {},
};
