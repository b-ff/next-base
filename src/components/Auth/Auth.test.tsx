import {
  SessionContext,
  SessionContextValue,
  signIn,
  signOut,
  useSession,
} from "next-auth/react";
import { Auth } from "./Auth";
import { render } from "@testing-library/react";

const mockUser = {
  email: "kermit@muppets.com",
  name: "Kermit The Frog",
  image:
    "https://lumiere-a.akamaihd.net/v1/images/character_themuppets_kermit_b77a431b.jpeg",
};

const mockSession = {
  update: () => Promise.resolve({}),
  data: { user: mockUser, expires: "" },
  status: "authenticated",
};

describe("Component Auth", () => {
  it("Should render as expected", () => {
    const { container } = render(<Auth />);
    expect(container).toMatchSnapshot();
  });

  it("Should call useSession hook from next-auth/react module", () => {
    const dom = render(<Auth />);
    expect(useSession).toBeCalled();
  });

  it("Should show sign-in link if user is not signed in", () => {
    const dom = render(<Auth />);
    expect(dom.getByText("Sign In / Sign Up")).not.toBeNull();
  });

  it("Should call signIn function from next-auth/react by clicking sign-in link", () => {
    const dom = render(<Auth />);
    dom.getByText("Sign In / Sign Up").click();
    expect(signIn).toBeCalled();
  });

  it("Should show sign-out link if user is signed in", () => {
    const dom = render(
      <SessionContext.Provider value={mockSession as SessionContextValue}>
        <Auth />
      </SessionContext.Provider>
    );

    expect(dom.getByText("Sign Out")).not.toBeNull();
  });

  it("Should call signOut function from next-auth/react by clicking sign-out link", () => {
    const dom = render(
      <SessionContext.Provider value={mockSession as SessionContextValue}>
        <Auth />
      </SessionContext.Provider>
    );

    dom.getByText("Sign Out").click();
    expect(signOut).toBeCalled();
  });
});
