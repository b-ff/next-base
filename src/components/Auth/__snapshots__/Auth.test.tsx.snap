// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component Auth Should render as expected 1`] = `
<div>
  <a
    href="#"
  >
    Sign In / Sign Up
  </a>
</div>
`;
